### ISSUE TITLE
`nixosTest`: How to get `stdout`/`stderr` as they come during a `execute` or `succeed`?

### ISSUE THREAD

**Issue by sir4ur0n** [NixOS/nixpkgs#219352]
I have started using `nixosTest` (which is great!) for integration tests in a project, and so far I have not found a way to print the `stdout/stderr` as they come while using `execute` or `succeed`. Printing `stdout` once the command is done is not good enough.

My use case:
* the integration tests can take time to run (tens of minutes).
* there is no way to know if a test is stuck because nothing is printing until it's done
* even then, there is no way to see what time was spent where, as everything is printed at the end

Is this possible?
* If not, is it possible to support this use case?
* If yes, can we please document it in https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests?

cc @roberth who seemed active on `nixosTest` issues in the past :sweat_smile:

**Comment by symphorien** [comment]
a possible "hack" is to pipe the executed command to `logger`.

**Comment by AleXoundOS** [comment]
Something like this works (but ugly):
```python
client.succeed("curl --fail http://server/ >&2")
```

**Comment by roberth** [comment]
> but ugly

It's the way we do it though.

I suppose it's unexpected that succeed returns its stdout as a python string. We could add a method that only executes and checks the exit code, and then we could rename `succeed` to `readStdout`, with a deprecation warning that mentions both new method names.

**Comment by sir4ur0n** [comment]
Thank you @AleXoundOS that does the trick indeed! :heart: 

It might be worth adding to the documentation, what do you all think? The only occurrence of `>&2` is in:
> If the command detaches, it must close stdout, as execute will wait for this to consume all output reliably. This can be achieved by redirecting stdout to stderr >&2, to /dev/console, /dev/null or a file.

but I read that several times and it wasn't clear it applied in my case. It's not that my command detaches, it's that I want stdout streamed, not dumped at the end :sweat_smile:

**Comment by AleXoundOS** [comment]
What I do not like about `>&2` is that this is a change to the command itself, rather than outputs routing on the test-driver side.

> rename succeed to readStdout

The first idea that came to mind about improving the `nixosTest` interface was to add an optional parameter to `succeed` function, to control `stdout`/`stderr` outputs routing.

`readStdout` may be indeed closer to conventional programming. But on the other hand, we need to think twice before introducing more methods and deprecating something...

**Comment by roberth** [comment]
cc @tfc

**Comment by tfc** [comment]
> `readStdout` may be indeed closer to conventional programming. But on the other hand, we need to think twice before introducing more methods and deprecating something...

Yes, exactly. If we add every use case to please everyone, the whole driver will become completely unmaintainable in a few years due to featuritis.
I'd prefer to not increase the number of methods and parameters on the test driver and have people build their own helpers on top of it. I think the goal of the test driver should be to provide the minimal set of methods that are as policy-free as possible to users who use their own helper functions to provide higher-level abstractions. We didn't have a good start here because the driver was already very rich in methods when I ported it from perl to python (couldn't just drop all that functionality in the existing >300 tests).

Looking at the functionality that @sir4ur0n is asking for, i have questions: Do you run another test framework in the integration test driver?
It seems like you tried to build this using the methods (`succeed` etc.) that are meant to represent test cases and not whole tests. 
It feels like you would want to create some kind of class that gets a process running in the background and connects the outputs to sockets/files/TCP streams and watches them. I would argue that that is functionality that can be on top of the driver, not inside the driver. Does this make sense?
I'd be always happy to help, maybe in some chat about design constraints.

**Comment by sir4ur0n** [comment]
> Do you run another test framework in the integration test driver?

Yes ([Cypress tests](https://www.cypress.io/) if you are curious).

> It seems like you tried to build this using the methods (succeed etc.) that are meant to represent test cases and not whole tests.

The way I see it (keeping in mind that I am rather new to `nixosTest` :sweat_smile: ), `nixosTest` contains several features:
1. describe an infrastructure (several NixOS VMs, a.k.a. nodes)
2. create that infrastructure (pop the well-configured QEMU VMs) and destroy it at the end
3. infrastructure API: `start_all()`, `wait_XXX()` family of functions, etc.
4. test API: `succeed()`, `execute()`, etc.

I use `nixosTest` for 1, 2 and 3, but I don't really care for 4. I want to call whatever builder/test framework I want instead on a particular VM, see the output as it comes, and fail if the return code is not `0`.

If I am misusing `nixosTest` (as in, "great if it works for your use case, but we don't actively support that"), then I guess we can close this issue; I will use the `>&2` trick and hope it does not break in future updates :smile: 

> It feels like you would want to create some kind of class that gets a process running in the background and connects the outputs to sockets/files/TCP streams and watches them. I would argue that that is functionality that can be on top of the driver, not inside the driver. Does this make sense?

If you are talking about creating a Python class and whatnot, I definitely don't want to do that in my repo, as much as possible :sweat_smile: I just want to run shell commands that trigger my test suite; I have no opinion on **how** this would be implemented though :neutral_face:

**Comment by roberth** [comment]
It would be nice to have a method or parameter that doesn't print a python stack trace when the test suite fails. For example, all of these lines are useless:

```
api: output: 
cleanup
kill machine (pid 8)
agent # qemu-kvm: terminating on signal 15 from pid 6 (/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/bin/python3.10)
kill machine (pid 19)
api # qemu-kvm: terminating on signal 15 from pid 6 (/nix/store/iw1vmh509hcbby8dbpsaanbri4zsq7dj-python3-3.10.10/bin/python3.10)
(finished: cleanup, in 0.05 seconds)
Traceback (most recent call last):
  File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/__init__.py", line 112, in main
    driver.run_tests()
  File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 142, in run_tests
    self.test_script()
  File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 138, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 11, in <module>
  File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 602, in succeed
    raise Exception(f"command `{command}` failed (exit code {status})")
Exception: command `(cd /nix/store/wcixwqqbcb5098cvx5hqizkznp3v8a8j-testdata && hercules-ci-agent-test  >/dev/console 2>/dev/console)` failed (exit code 1)
kill vlan (pid 7)
builder for '/nix/store/0q1dnycvc4kwfyig0cc2irlz8c4b9pm1-vm-test-run-agent-test.drv' failed with exit code 1;
last 10 log lines:
>     driver.run_tests()
>   File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 142, in run_tests
>     self.test_script()
>   File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 138, in test_script
>     exec(self.tests, symbols, None)
>   File "<string>", line 11, in <module>
>   File "/nix/store/vk10lxyq8b9c2hhv9wflqcmvc5g5z9sr-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 602, in succeed
>     raise Exception(f"command `{command}` failed (exit code {status})")
> Exception: command `(cd /nix/store/wcixwqqbcb5098cvx5hqizkznp3v8a8j-testdata && hercules-ci-agent-test  >/dev/console 2>/dev/console)` failed (exit code 1)
> kill vlan (pid 7)
For full logs, run 'nix log /nix/store/0q1dnycvc4kwfyig0cc2irlz8c4b9pm1-vm-test-run-agent-test.drv'.
```
